ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -717,6 +725,27 @@ private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws Sto
         IndexValue.Flags.Delete_Index));
   }
 
+  /**
+   * Calculate the farthest index that can be used for a bundle IO read based on start index and bundleReadBuffer capacity.
+   * @param sortedSrcIndexEntries all available entries, which are ordered by offset.
+   * @param start the staring index for the given list, not {@link IndexEntry}.
+   * @return the farthest index, inclusively, which can be used for a bundle read.
+   */
+  private int getBundleReadEndIndex(List<IndexEntry> sortedSrcIndexEntries, int start) {
+    long startOffset = sortedSrcIndexEntries.get(start).getValue().getOffset().getOffset();
+    int end;
+    for (end = start; end < sortedSrcIndexEntries.size(); end++) {
+      long currentSize =
+          sortedSrcIndexEntries.get(end).getValue().getOffset().getOffset() + sortedSrcIndexEntries.get(end)
+              .getValue()
+              .getSize() - startOffset;
+      if (currentSize > bundleReadBuffer.capacity()) {
+        break;
+      }
+    }
+    return end - 1;
+  }
```

Comment:
The java streams way of achieving this looks to be the following (just for information as some might find it more readable and clean - no need to make the change):
```java
 return IntStream.range(0, sortedSrcIndexEntries.size())
        .filter(i -> sortedSrcIndexEntries.get(i).getValue().getOffset().getOffset() + 
                     sortedSrcIndexEntries.get(i).getValue().getSize() - startOffset > 
                     bundleReadBuffer.capacity())
        .findFirst()
        .orElse(sortedSrcIndexEntries.size()) - 1;
```